File: dotnet/Lexy/Lexy.Compiler/Language/Expressions/Functions/NoArgumentFunction.cs
using System.Collections.Generic;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language.Expressions.Functions
{
    public abstract class NoArgumentFunction : BuiltInFunction
    {
        protected abstract VariableType ResultType { get; }

        protected NoArgumentFunction(SourceReference reference)
            : base(reference)
        {
        }

        public override IEnumerable<INode> GetChildren()
        {
            yield break;
        }

        protected override void Validate(IValidationContext context)
        {
        }

        public override VariableType DeriveReturnType(IValidationContext context) => ResultType;
    }
}

File: typescipt/src/language/expressions/functions/noArgumentFunction.ts


export class NoArgumentFunction extends ExpressionFunction {
   protected abstract VariableType ResultType

   constructor(reference: SourceReference)
     {
     super(reference);
   }

   public override getChildren(): Array<INode> {
     yield break;
   }

   protected override validate(context: IValidationContext): void {
   }

   public override deriveReturnType(context: IValidationContext): VariableType {
     return ResultType;
   }
}
